src/BlockTypes/ProductQuery.php

```diff
@@ -80,41 +83,111 @@ public function get_query_by_attributes( $query ) {
 			'posts_per_page' => $query['posts_per_page'],
 			'orderby'        => $query['orderby'],
 			'order'          => $query['order'],
+			'offset'         => $query['offset'],
+			// Ignoring the warning of not using meta queries.
+			// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_query
+			'meta_query'     => array(),
+		);
+
+		$queries_attributes = $this->get_queries_by_attributes( $variation_props );
+		$queries_filters    = $this->get_queries_by_applied_filters();
+
+		return array_reduce(
+			array_merge(
+				$queries_attributes,
+				$queries_filters
+			),
+			function( $acc, $query ) {
+				if ( isset( $query['post__in'] ) ) {
+					$acc['post__in'] = isset( $acc['post__in'] ) ? array_merge( $acc['post__in'], $query['post__in'] ) : $query['post__in'];
+				}
```

Comment:
If I'm understanding this correctly, I think this is working now because we only set the `post__in` attribute with the _on sale_ filter. But if at some point there were two filters using `post__in`, instead of concatenating the ids we should only include the intersection of `$acc['post__in']` and `$query['post__in']`, in other words, ids which are in both arrays so it behaves as an AND instead of an OR.